amethyst_ui/src/button/builder.rs

```diff
@@ -9,43 +9,47 @@ use amethyst_core::{
 use amethyst_renderer::{Texture, TextureHandle};
 
 use crate::{
-    font::default::get_default_font, Anchor, FontAsset, FontHandle, MouseReactive, OnUiActionImage,
-    OnUiActionSound, Stretch, UiButton, UiImage, UiText, UiTransform,
+    font::default::get_default_font, Anchor, FontAsset, FontHandle, Interactable, OnUiActionImage,
+    OnUiActionSound, Selectable, Stretch, UiButton, UiText, UiTransform,
 };
 
+use std::marker::PhantomData;
+
 const DEFAULT_Z: f32 = 1.0;
 const DEFAULT_WIDTH: f32 = 128.0;
 const DEFAULT_HEIGHT: f32 = 64.0;
-const DEFAULT_TAB_ORDER: i32 = 9;
+const DEFAULT_TAB_ORDER: u32 = 9;
 const DEFAULT_BKGD_COLOR: [f32; 4] = [0.82, 0.83, 0.83, 1.0];
 const DEFAULT_TXT_COLOR: [f32; 4] = [0.0, 0.0, 0.0, 1.0];
 
 /// Container for all the resources the builder needs to make a new UiButton.
 #[derive(SystemData)]
-pub struct UiButtonBuilderResources<'a> {
+pub struct UiButtonBuilderResources<'a, G: PartialEq + Send + Sync + 'static> {
     font_asset: Read<'a, AssetStorage<FontAsset>>,
     texture_asset: Read<'a, AssetStorage<Texture>>,
     loader: ReadExpect<'a, Loader>,
     entities: Entities<'a>,
-    image: WriteStorage<'a, UiImage>,
-    mouse_reactive: WriteStorage<'a, MouseReactive>,
+    image: WriteStorage<'a, TextureHandle>,
+    mouse_reactive: WriteStorage<'a, Interactable>,
     parent: WriteStorage<'a, Parent>,
     text: WriteStorage<'a, UiText>,
     transform: WriteStorage<'a, UiTransform>,
     button: WriteStorage<'a, UiButton>,
     action_image: WriteStorage<'a, OnUiActionImage>,
     action_sound: WriteStorage<'a, OnUiActionSound>,
+    selectables: WriteStorage<'a, Selectable<G>>,
 }
 
 /// Convenience structure for building a button
-pub struct UiButtonBuilder {
+#[derive(Debug, Clone)]
+pub struct UiButtonBuilder<G> {
     name: String,
     x: f32,
     y: f32,
     z: f32,
     width: f32,
     height: f32,
-    tab_order: i32,
+    tab_order: u32,
```

Comment:
Allowing these to be negative can make quick UI edits a bit easier if say I want to add a control to the top of a mostly completed form.  If I can't use negatives then I'd have to change the tab order in everything else too.

Reply:
From a logical perspective, it doesn't make sense ^^

I agree it does make prototyping easier, but you can fix that with:
1. Automatic tab_order generation
2. Partitioning your tab_order properly (leave gaps between elements, don't start at 0)
3. Planning your UI in advance

Reply:
That's true. I just changed it to u32 because its hard to reason about what a tab_order of -10 means in reality. Its not intuitive.

Also a lot of people seemed to think a tab_order of -1 meant disabled, when in reality it wasn't the case.

Reply:
Up to you. I don't think it will matter when we have a prefab editor, because we can totally automate this. We can have a "increment all" button, or a "resolve conflicts" button which works a bit like when you resolve merge conflicts, asking you which one you want first.

Reply:
Huh those are good points.  Let's leave this as is then.